src/grammar.adoc

```diff
@@ -0,0 +1,208 @@
+
+|===
+
+
+==== Catalog Access
+
+[{bnf}]
+|===
+|load-decl       |::=|  `(` `*load*` _load-body_ `)`
+|import-decl     |::=|  `(` `*import*` _load-body_ `)`
+|load-body       |::=| _module-name_ _catalog-name_ _catalog-version_ _symbol-maxid_?
+
+|catalog-name    |::=|  _unannotated-string_
+|catalog-version |::=|  _unannotated-uint_
+|symbol-maxid    |::=|  _unannotated-uint_
+|===
+
+
+=== Macro References
+
+[{bnf}]
+|===
+|macro-ref     |::=|  _macro-name_  \|  _local-ref_  \|  _qualified-ref_
+
+|local-ref     |::=|  < symbol of the form ``':``_name-or-address_``'`` >
+|qualified-ref |::=|  < symbol of the form ``':``_module-name_``:``_name-or-address_``'`` >
+
+|module-name     |::=|  _unannotated-identifier-symbol_
+|macro-name      |::=|  _unannotated-identifier-symbol_
+|macro-address   |::=|  _unannotated-uint_
+|name-or-address |::=|  _macro-name_  \|  _macro-address_
+|===
+
+
+=== Module Definitions
+
+[{bnf}]
+|===
+|inline-module-def   |::=| `(` `*module*` _module-name_ _module-body_ `)`
+|shared-module-def   |::=| ``$ion_shared_module::``__ion-version-marker__``::(`` _catalog-key_
+_module-body_ `)`
+|catalog-key         |::=| `(` `*catalog_key*` _catalog-name_ _catalog-version_ `)`
+|===
+
+
+==== Module Bodies
+
+[{bnf}]
+|===
+|module-body |::=|  _dependency_* _symtab_? _macro-alias_* _module-mactab_?
+
+|macro-alias |::=|  `(` `*alias*` _macro-name_ _macro-ref_ `)`
+
+|module-mactab   |::=|  `(` `*macro_table*` _macro-or-export_* `)`
+|macro-or-export |::=|  _macro-defn_  \|  _export_
+
+|export      |::=|  `(` `*export*` _export-item_* `)`
+|export-item |::=|  _macro-ref_                                         +
+                \|  `(` `*from*` _module-name_ _name-or-address_* `)`   +
+                \|  `(` `*all_from*` _module-name_ `)`
+|===
+
+
+=== Macro Definitions
+
+[{bnf}]
+|===
+|macro-defn |::=|  `(` `*macro*` _macro-name_? _signature_ _template_ `)`
+
+|signature   |::=|  _param-specs_ _result-spec_?
+|param-specs |::=|  `(` _param-spec_* _rest-spec_? `)`  \|  `[` _param-spec_* _rest-spec_? `]`
+|param-spec  |::=|  _param-name_  \|  _simple-param-spec_  \|  _grouped-param-spec_
+|param-name  |::=|  _unannotated-identifier-symbol_
+
+|simple-param-spec  |::=|  `(` _param-name_  _base-shape_ _simple-cardinality_ `)`
+|base-shape         |::=| _tagged-type_  \|  _primitive-type_  \|  _macro-ref_
+|simple-cardinality |::=|  `*!*`  \|  `*+*`  \|  `*'?'*`  \|  `*'{asterisk}'*`
+
+|grouped-param-spec  |::=|  `(` _param-name_ `[` _base-shape_ `]` _grouped-cardinality_? `)`
+|grouped-cardinality |::=|  `*+*`  \|  `*'{asterisk}'*`
+
+|rest-spec        |::=|  `(` _param-name_ _base-shape_ _rest-cardinality_  `)`
+|rest-cardinality |::=|   `*\...*`  \|  `*\...+*`
+
+|tagged-type   |::=| _abstract-type_ \| _concrete-type_
+|abstract-type |::=| `*any*` \| `*number*` \| `*exact*` \| `*text*` \| `*lob*` \| `*sequence*`
+|concrete-type |::=|  `*'null'*`  \|  `*bool*`  \|  `*timestamp*`  \|  `*int*`  \|  `*decimal*`
```

Comment:
Are *`null`* and *`'null'`* are allowed as syntactic synonyms?  And if *`null`* works, presumably *`null.null`* would work too.

Reply:
Prior revisions allowed `null` but everyone seems unhappy with that.  Here, these are all actual symbols (thus the `'`s), so `'null'` is correct but `null` and `null.null` are erroneous.